File: src/python_pkg/python_pkg/adaptive_resolution_node.py
import numbers

from numpy import size
import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry
from rcl_interfaces.msg import Parameter, ParameterType, ParameterValue
from rcl_interfaces.srv import SetParameters
import math

class AdaptiveResolutionNode(Node):
    def __init__(self):
        super().__init__('adaptive_resolution_node')
        
        # Parameters
        self.declare_parameter('low_speed_threshold', 0.2)
        self.declare_parameter('high_speed_threshold', 0.3)
        self.declare_parameter('enable_dynamic_decimation', True)
        
        self.low_speed_limit = self.get_parameter('low_speed_threshold').value
        self.high_speed_limit = self.get_parameter('high_speed_threshold').value
        
        # State
        self.current_state = 'HIGH_RES' # HIGH_RES (mag 1) or LOW_RES (mag 2/3)
        
        # Camera Node Name (hardcoded to default official node name, adjust if needed)
        self.camera_node_name = '/camera/camera'
        
        # Subscriber
        self.odom_sub = self.create_subscription(
            Odometry,
            '/odom',
            self.odom_callback,
            10
        )
        
        # Service Client for Parameter Setting
        # 1. Setup (The "Phone Call")
        # The node prepares a "phone line" (Service Client) to talk to the RealSense camera node.
        # Specifically, it wants to talk to the /camera/camera/set_parameters service, which is the standard ROS 2 way to change settings of another node remotely.

        self.param_client = self.create_client(SetParameters, f'{self.camera_node_name}/set_parameters')
        
        self.get_logger().info('Adaptive Resolution Node Started')
        self.get_logger().info(f'Waiting for {self.camera_node_name}/set_parameters service...')
        self.param_client.wait_for_service(timeout_sec=5.0)
        
    def odom_callback(self, msg):
        if not self.get_parameter('enable_dynamic_decimation').value:
            return

        #1. Calculate linear speed
        # it listens to /odom and calcualtes linear speed as sqrt(vx^2 + vy^2). If the speed exceeds the high threshold, it calls the set_decimation function to switch to low resolution (decimation 2). If the speed drops below the low threshold, it switches back to high resolution (decimation 1). The hysteresis prevents rapid toggling between resolutions when the speed is around the threshold.
        vx = msg.twist.twist.linear.x
        vy = msg.twist.twist.linear.y
        speed = math.sqrt(vx**2 + vy**2)
        
        #3. Hysteresis Logic to prevent flickering
        # IF you are driving FAST (> 0.3 m/s), it switches to "Low Res".
        # IF you slow down BELOW 0.2 m/s, it switches back to "High Res".
        # Why two numbers (0.3 and 0.2)? This is called Hysteresis. If we just used one number (e.g., 0.25), and you drove at exactly 0.25 m/s, the camera would glitch out switching between High/Low resolution 10 times a second. This gap prevents that.
        
        if self.current_state == 'HIGH_RES' and speed > self.high_speed_limit:
            self.set_decimation(2) # Low Resolution (Half size)
            self.current_state = 'LOW_RES'
            self.get_logger().info(f'Speed {speed:.2f} > {self.high_speed_limit}: Switching to Low Res (Decimation 2)')
            
        elif self.current_state == 'LOW_RES' and speed < self.low_speed_limit:
            self.set_decimation(1) # High Resolution (Original size)
            self.current_state = 'HIGH_RES'
            self.get_logger().info(f'Speed {speed:.2f} < {self.low_speed_limit}: Switching to High Res (Decimation 1)')

    def set_decimation(self, magnitude):
        if not self.param_client.service_is_ready():
            return
            
        req = SetParameters.Request()
        val = ParameterValue(type=ParameterType.PARAMETER_INTEGER, integer_value=magnitude)
        param = Parameter(name='decimation_filter.filter_magnitude', value=val)
        req.parameters = [param]
        # 4. Changing the Setting 
        # This is where it actually "makes the call" to the RealSense node.
        # It tells the camera: "Hey, set your decimation_filter.filter_magnitude to X".
        # Magnitude 1: Keep original size (424x240).
        # Magnitude 2: Divide size by 2 (212x120) -> 4x less data!
        future = self.param_client.call_async(req)
        # We don't block waiting for result to keep odom callback fast

def main(args=None):
    rclpy.init(args=args)
    node = AdaptiveResolutionNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/jetcar_common/src/adaptive_resolution_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "rcl_interfaces/srv/set_parameters.hpp"
#include "rcl_interfaces/msg/parameter.hpp"
#include "rcl_interfaces/msg/parameter_value.hpp"
#include "rcl_interfaces/msg/parameter_type.hpp"

#include <cmath>
#include <chrono>

using namespace std::chrono_literals;

class AdaptiveResolutionNode : public rclcpp::Node
{
public:
  AdaptiveResolutionNode()
  : Node("adaptive_resolution_node")
  {
    // Parameters
    this->declare_parameter("low_speed_threshold", 0.2);
    this->declare_parameter("high_speed_threshold", 0.3);
    this->declare_parameter("enable_dynamic_decimation", true);
    this->declare_parameter("camera_node_name", "/camera/camera");

    low_speed_limit_ = this->get_parameter("low_speed_threshold").as_double();
    high_speed_limit_ = this->get_parameter("high_speed_threshold").as_double();
    camera_node_name_ = this->get_parameter("camera_node_name").as_string();

    current_state_ = HIGH_RES;

    // Subscriber
    odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
      "/odom", 10, std::bind(&AdaptiveResolutionNode::odom_callback, this, std::placeholders::_1));

    // Service Client
    std::string service_name = camera_node_name_ + "/set_parameters";
    param_client_ = this->create_client<rcl_interfaces::srv::SetParameters>(service_name);

    RCLCPP_INFO(this->get_logger(), "Adaptive Resolution Node Started");
    RCLCPP_INFO(this->get_logger(), "Waiting for %s service...", service_name.c_str());
    
    // We won't block here with wait_for_service inside constructor to avoid blocking the node initialization
    // The service check will happen before calling it.
  }

private:
  enum State {
    HIGH_RES,
    LOW_RES
  };

  void odom_callback(const nav_msgs::msg::Odometry::SharedPtr msg)
  {
    if (!this->get_parameter("enable_dynamic_decimation").as_bool()) {
      return;
    }

    double vx = msg->twist.twist.linear.x;
    double vy = msg->twist.twist.linear.y;
    double speed = std::sqrt(vx * vx + vy * vy);

    if (current_state_ == HIGH_RES && speed > high_speed_limit_) {
      set_decimation(2); // Low Resolution
      current_state_ = LOW_RES;
      // RCLCPP_INFO(this->get_logger(), "Speed %.2f > %.2f: Switching to Low Res (Decimation 2)", speed, high_speed_limit_);
    } else if (current_state_ == LOW_RES && speed < low_speed_limit_) {
      set_decimation(1); // High Resolution
      current_state_ = HIGH_RES;
      // RCLCPP_INFO(this->get_logger(), "Speed %.2f < %.2f: Switching to High Res (Decimation 1)", speed, low_speed_limit_);
    }
  }

  void set_decimation(int magnitude)
  {
    if (!param_client_->service_is_ready()) {
      // Just return silently to avoid log spam, mirroring python behavior which checks is_ready
      return;
    }

    auto request = std::make_shared<rcl_interfaces::srv::SetParameters::Request>();
    rcl_interfaces::msg::Parameter param;
    param.name = "decimation_filter.filter_magnitude";
    param.value.type = rcl_interfaces::msg::ParameterType::PARAMETER_INTEGER;
    param.value.integer_value = magnitude;
    request->parameters.push_back(param);
    
    // Asynchronous call
    param_client_->async_send_request(request);
  }

  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
  rclcpp::Client<rcl_interfaces::srv::SetParameters>::SharedPtr param_client_;
  
  double low_speed_limit_;
  double high_speed_limit_;
  std::string camera_node_name_;
  State current_state_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<AdaptiveResolutionNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
